### PR TITLE
Rule Based Query Optimizer

### PR DESCRIPTION
**Pull request by kcaras**
Here is a Rule Based Query Optimizer that implements predicate pushdown, projection pushdown, join elimination, and simplify predicates such as 1==0. 

There are a total of 10 tests.

### CODE REVIEW

**Comment by gaurav274** [comment]
Can you please make changes we discussed?
1. Move tests to tests dir
2. Remove all the binary files (*.avi files)
3. Remove the spatial and temporal filtering files. We can take them as a separate PR.

File: src/query_planner/abstract_select_plan.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
+"""Abstract class for all the scan planners
```
Comment (gaurav274) [inline src/query_planner/abstract_select_plan.py]: Confused. we already have an abstract scan then why abstract select?

File: src/query_planner/abstract_table_plan.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
+"""Abstract class for all the scan planners
```
Comment (gaurav274) [inline src/query_planner/abstract_table_plan.py]: What is a table plan?

File: src/query_planner/video_table_plan.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
+from src.loaders.abstract_loader import AbstractVideoLoader
```
Comment (gaurav274) [inline src/query_planner/video_table_plan.py]: This shouldn't be a plan node right?

File: src/query_optimizer/statement2plantree.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+from models.catalog.video_info import VideoMetaInfo
+from models.catalog.properties import VideoFormat
+
+
+# class that will handle converting a statement to a plan tree
+class Statement2Plantree:
+    def __init__(self):
+        pass
+
+    # Function that converts the statement list to a plan tree
```
Comment (gaurav274) [inline src/query_optimizer/statement2plantree.py]: Comments are not as per our standards. And why is there a class for just one static function? We need to keep some sort of framework where other modules just call a single function with evaStatementList and then it is the job of this module to figure out which type of statement it is and call the corresponding converter.

File: src/query_optimizer/rule_query_optimizer_playground.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
+# this is an implementation of a Rule Based Query Optimizer
```
Comment (gaurav274) [inline src/query_optimizer/rule_query_optimizer_playground.py]: Remove this?

File: src/query_optimizer/rule_query_optimizer.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
+from expression.constant_value_expression import ConstantValueExpression
```
Comment (gaurav274) [inline src/query_optimizer/rule_query_optimizer.py]: Can you please modify the commenting style? Use docstring. I will look at the functionality once you are done with this.

File: src/query_optimizer/rule_query_optimizer_playground.py (hunk 5 above)
Comment (bkim360) [reply src/query_optimizer/rule_query_optimizer_playground.py]: Done.

File: src/query_planner/abstract_select_plan.py (hunk 1 above)
Comment (bkim360) [reply src/query_planner/abstract_select_plan.py]: We were not sure about the naming scheme. Should we just port over the methods from abstract select to abstract scan and delete abstract select after?

File: src/query_planner/abstract_table_plan.py (hunk 2 above)
Comment (bkim360) [reply src/query_planner/abstract_table_plan.py]: We thought we would need a table plan for the videos (the table representing the video).

File: src/query_optimizer/rule_query_optimizer.py (hunk 6 above)
Comment (bkim360) [reply src/query_optimizer/rule_query_optimizer.py]: Changed them for rule_query_optimizer.py and the abstract plans.

File: src/query_planner/abstract_select_plan.py (hunk 1 above)
Comment (gaurav274) [reply src/query_planner/abstract_select_plan.py]: Yeah go ahead and do that.

File: src/query_planner/abstract_table_plan.py (hunk 2 above)
Comment (gaurav274) [reply src/query_planner/abstract_table_plan.py]: Have a look at the table_ref in src/query_parser/table_ref.py.

Comment (bkim360) [reply src/query_planner/abstract_table_plan.py]: Does this mean we should scrap the video_table_plan and abstract_table_plan and replace them with the table_ref?

File: src/query_planner/abstract_select_plan.py (hunk 1 above)
Comment (bkim360) [reply src/query_planner/abstract_select_plan.py]: I have changed the references of select plan to seq scan plan. If this is correct, I will delete the abstract select plan and logical select plan.

File: src/query_planner/abstract_table_plan.py (hunk 2 above)
Comment (bkim360) [reply src/query_planner/abstract_table_plan.py]: I have changed the references of video_table_plan and abstract_table_plan into table_ref.

File: src/query_planner/video_table_plan.py (hunk 3 above)
Comment (bkim360) [reply src/query_planner/video_table_plan.py]: Mentioned above.

**Pull request closed without merging**